homeassistant/components/mqtt/discovery.py

```diff
@@ -239,59 +365,125 @@ def async_discovery_message_received(msg: ReceiveMessage) -> None:  # noqa: C901
+                )
+                discovered_components.append(
+                    MqttComponentConfig(
+                        component,
+                        component_object_id,
+                        component_node_id,
+                        discovery_payload,
+                    )
+                )
+            _LOGGER.debug(
+                "Process device discovery payload %s", device_discovery_payload
+            )
+            device_discovery_id = f"{node_id} {object_id}" if node_id else object_id
+            message = f"Processing device discovery for '{device_discovery_id}'"
+            async_log_discovery_origin_info(
+                message, MQTTDiscoveryPayload(device_discovery_payload)
+            )
+
+        else:
+            # Process component based discovery message
             try:
-                discovery_payload = MQTTDiscoveryPayload(json_loads_object(payload))
+                discovery_payload = MQTTDiscoveryPayload(
+                    json_loads_object(payload) if payload else {}
+                )
             except ValueError:
                 _LOGGER.warning("Unable to parse JSON %s: '%s'", object_id, payload)
                 return
-            _replace_all_abbreviations(discovery_payload)
-            if not _valid_origin_info(discovery_payload):
-                return
+            if not _async_process_discovery_migration(discovery_payload):
+                _replace_all_abbreviations(discovery_payload)
+                if not _valid_origin_info(discovery_payload):
+                    return
+            discovered_components.append(
+                MqttComponentConfig(component, object_id, node_id, discovery_payload)
+            )
+
+        discovery_pending_discovered = mqtt_data.discovery_pending_discovered
+        for component_config in discovered_components:
+            component = component_config.component
+            node_id = component_config.node_id
+            object_id = component_config.object_id
+            discovery_payload = component_config.discovery_payload
+
             if TOPIC_BASE in discovery_payload:
                 _replace_topic_base(discovery_payload)
-        else:
-            discovery_payload = MQTTDiscoveryPayload({})
 
-        # If present, the node_id will be included in the discovered object id
-        discovery_id = f"{node_id} {object_id}" if node_id else object_id
-        discovery_hash = (component, discovery_id)
+            # If present, the node_id will be included in the discovery_id.
+            discovery_id = f"{node_id} {object_id}" if node_id else object_id
+            discovery_hash = (component, discovery_id)
 
-        if discovery_payload:
             # Attach MQTT topic to the payload, used for debug prints
-            setattr(
-                discovery_payload,
-                "__configuration_source__",
-                f"MQTT (topic: '{topic}')",
-            )
```

Comment:
Maybe I've already asked, but why is this removed? Was it never used?

Reply:
It was never used.